CSharp/Library/Microsoft.Bot.Connector.Shared/Authentication/AdalAuthenticator.cs

```diff
@@ -0,0 +1,115 @@
+﻿using Microsoft.Extensions.Logging;
+using Microsoft.IdentityModel.Clients.ActiveDirectory;
+using System;
+using System.Collections.Generic;
+using System.Net.Http.Headers;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Bot.Connector.Shared.Authentication
+{
+    public class AdalAuthenticator 
+    {
+        private static readonly TimeSpan SemaphoreTimeout = TimeSpan.FromSeconds(1);
+        private const int MaxRetries = 5;
+
+        private static AuthenticationContext authContext = new AuthenticationContext(JwtConfig.ToChannelFromBotLoginUrl) { ExtendedLifeTimeEnabled = true };
+        private static SemaphoreSlim authContextSemaphore = new SemaphoreSlim(50, 50);
+
+        private readonly ClientCredential clientCredential;
+
+        public AdalAuthenticator(ClientCredential clientCredential)
+        {
+            this.clientCredential = clientCredential ?? throw new ArgumentNullException(nameof(clientCredential));
+        }
+
+        public async Task<AuthenticationResult> GetTokenAsync()
+        {
+            return await Retry.Run(
+                task: () => AcquireTokenAsync(),
+                retryExceptionHandler: (ex, ct) => HandleAdalException(ex, ct)).ConfigureAwait(false);
+        }
+
+        public void ClearTokenCache()
+        {
+            // To force a refresh, just clear the AuthenticationContext's internal token cache.
+            authContext.TokenCache.Clear();
+        }
+
+        private async Task<AuthenticationResult> AcquireTokenAsync()
+        {
+            bool acquired = false;
+
+            try
+            {
+                // The ADAL client team recommends limiting concurrency of calls. When the Token is in cache there is never 
+                // contention on this semaphore, but when tokens expire there is some. However, after measuring performance
+                // with and without the semaphore (and different configs for the semaphore), not limiting concurrency actually
+                // results in higher response times overall. Without the use of this semaphore calls to AcquireTokenAsync can take up
+                // to 5 seconds under high concurrency scenarios.
+                acquired = await authContextSemaphore.WaitAsync(SemaphoreTimeout).ConfigureAwait(false);
```

Comment:
The semaphore acts as a concurrency limiter for ADAL. Yes, in practice, there is contention only when a token renewal is underway.

Note that if we don't acquire the semaphore we will retry anyways because we treat TimeOutException as a retryable exception, so we are actually enduring ~5 seconds of delay before returning an exception. So if the acquisition takes >= 1 second, there will be a timeout exception, a retry will be issued and then when it comes bakc the token will be acquired and there will be no contention on the semaphore.

An alternative implementation which I experimented on would to not retry 5 times to acquire the semaphore for 1 second, and just wait on the semaphore 5 seconds. We could consider that, at some point I experimented with it.

After some retrying / waiting though, I think that given this is a message in a conversation, if we don't achieve anything by 5 or 10 seconds we should just fail. 

I will swing by and talk about this one in person to make sure we land at the best decision possible here :)